source/extensions/filters/network/kafka/codec.cc

```diff
@@ -0,0 +1,52 @@
+#include "extensions/filters/network/kafka/codec.h"
+
+#include "extensions/filters/network/kafka/kafka_protocol.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+
+void RequestDecoder::onData(Buffer::Instance& data) {
+  uint64_t num_slices = data.getRawSlices(nullptr, 0);
+  Buffer::RawSlice slices[num_slices];
+  data.getRawSlices(slices, num_slices);
+  for (const Buffer::RawSlice& slice : slices) {
+    doParse(request_parser_, slice);
+  }
+}
+
+void RequestDecoder::doParse(ParserSharedPtr& parser, const Buffer::RawSlice& slice) {
+  const char* buffer = reinterpret_cast<const char*>(slice.mem_);
```

Comment:
basic parser loop:
`RequestStartParser` (consumes length) returns `RequestHeaderParser`
`RequestHeaderParser` (consumes 4 fields) returns type-specific `XParser`
`XParser` returns a message instead of parser
loop starts anew